# Seurat 4 源码阅读

```{r setup, include=FALSE}
# 源码解析中一大半代码是用来展示的，只有一小部分需要运行。
# 该文件包含 绪论, 

# todo: 
# 1,2,3
```

这几篇主要解读重要步骤的函数。分别面向3类读者，调包侠，R包写手，一般R用户。这也是我自己的三个身份。

- 调包侠关心生物学问题即可，比如数据到底怎么标准化的，是否scale过，怎么求平均表达量？
- R包写手则要关心更多细节，需要阅读源码及注释。
- 而一般R用户则可以直接看最后的R tips，学习R似乎无尽的函数和使用技巧，这是阅读源码学习大神内功的第二手资料。


解读目的：

- 美国科技就在手中，但是99%的时间只是用，想改发现太复杂，牵一发而动全身，就很不甘心。
- 天天开挖掘机，这一次算是把挖掘机拆开了，看看怎么造的。
- 向最好最流行的包学习，一次读不懂就多读几遍，温故而知新。
- 提升代码能力，写高质量代码，利国利民。
- 吹不动了，开始干活


> 锁定版本: seurat-4.1.0, seurat-object-4.0.4 截止 2022.1.31

> [知乎首发](https://zhuanlan.zhihu.com/p/465392721):https://zhuanlan.zhihu.com/p/465392721
> 喜欢可以点赞、关注、收藏。



我们如果想对Seurat R包的实现有更深入的理解，可以看看源码解析系列。

如果时间有限，可以只看每一篇的首尾2部分，第一部分讲了怎么用，第二部分是源码注释，第三部分是R tips。


> scRNAseq分析主要代码

应用层经常用到的就这几十行代码。里面的step1-12编号是我自己加的，方便定位和交流。


```
library(dplyr)
library(Seurat)
library(patchwork)


# load data
# step1 load df ====
pbmc.data <- Read10X(data.dir = "~/data/scScripts/backup/data/pbmc3k/filtered_gene_bc_matrices/hg19/")
class(pbmc.data)
dim(pbmc.data) #32738  2700


# Initialize the Seurat object with the raw (non-normalized data).
# step2 create Seurat obj====
pbmc <- CreateSeuratObject(counts = pbmc.data,
                           project = "pbmc3k",
                           min.cells = 3,
                           min.features = 200)
pbmc
pbmc@assays$RNA@scale.data #<0 x 0 matrix>



# step3 add meta.data
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
pbmc[["percent.rp"]] <- PercentageFeatureSet(pbmc, pattern = "^RP[SL]")

# cell cycle
pbmc <- CellCycleScoring(pbmc,
                         s.features = cc.genes$s.genes,
                         g2m.features = cc.genes$g2m.genes)



# step4 QC
# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rp"), ncol = 4, pt.size=0)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

# filter
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)




# step5 Normalizing the data
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)



# step6 Identification of highly variable features (feature selection)
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
#top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
#plot1 <- VariableFeaturePlot(pbmc)
#plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
#plot1 + plot2




# step7 Scaling the data
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)



# step8 (PCA) Perform linear dimensional reduction
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))

# Examine and visualize PCA results a few different ways
#print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
# VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
DimPlot(pbmc, reduction = "pca")

# DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
# DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)




# step9 Determine the 'dimensionality' of the dataset
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
#pbmc <- JackStraw(pbmc, num.replicate = 100)
#pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
# JackStrawPlot(pbmc, dims = 1:15)
ElbowPlot(pbmc)




# step10 Cluster the cells
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)

# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)




# step11 Run non-linear dimensional reduction (UMAP/tSNE)
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')
pbmc <- RunUMAP(pbmc, dims = 1:10)

# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap")

# saveRDS(pbmc, file = "../output/pbmc_tutorial.rds")




# step12 Finding differentially expressed features (cluster biomarkers)
# find all markers of cluster 2
#cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25)
#head(cluster2.markers, n = 5)


# find all markers distinguishing cluster 5 from clusters 0 and 3
#cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
#head(cluster5.markers, n = 5)


# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)  #可能有版本差异 avg_log2FC
# cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)

VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)

FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP",
                               "CD8A"))

# heatmap
top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(pbmc, features = top10$gene, label = T)  #+ NoLegend()






# step 13 Assigning cell type identity to clusters
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono",
                     "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()


# saveRDS(pbmc, file = "../output/pbmc3k_final.rds")
```




